python/lsst/ip/diffim/zogy.py

```diff
@@ -0,0 +1,839 @@
+        psf1 : 2D numpy.array
+            (Optional) Input psf of template, override if already padded
+        psf2 : 2D numpy.array
+            (Optional) Input psf of science image, override if already padded
+
+        Returns
+        -------
+        A lsst.pipe.base.Struct containing:
+        - Pr : 2D numpy.array, the (possibly zero-padded) template PSF
+        - Pn : 2D numpy.array, the (possibly zero-padded) science PSF
+        - Pr_hat : 2D numpy.array, the FFT of `Pr`
+        - Pn_hat : 2D numpy.array, the FFT of `Pn`
+        - denom : 2D numpy.array, the denominator of equation (13) in ZOGY (2016) manuscript
+        - Fd : float, the relative flux scaling factor between science and template
+        """
+        psf1 = self.im1_psf if psf1 is None else psf1
+        psf2 = self.im2_psf if psf2 is None else psf2
+        padSize = self.padSize if padSize is None else padSize
+        Pr, Pn = psf1, psf2
+        if padSize > 0:
+            Pr = self._padPsfToSize(psf1, (padSize, padSize))
+            Pn = self._padPsfToSize(psf2, (padSize, padSize))
+
+        sigR, sigN = self.sig1, self.sig2
+        Pr_hat = np.fft.fft2(Pr)
+        Pr_hat2 = np.conj(Pr_hat) * Pr_hat  # np.abs(Pr_hat)**2)
+        Pn_hat = np.fft.fft2(Pn)
+        Pn_hat2 = np.conj(Pn_hat) * Pn_hat  # np.abs(Pn_hat)**2))
+        denom = np.sqrt((sigN**2 * self.Fr**2 * Pr_hat2) + (sigR**2 * self.Fn**2 * Pn_hat2))
+        Fd = self.Fr*self.Fn / np.sqrt(sigN**2 * self.Fr**2 + sigR**2 * self.Fn**2)
+
+        res = pipeBase.Struct(
+            Pr=Pr, Pn=Pn, Pr_hat=Pr_hat, Pn_hat=Pn_hat, denom=denom, Fd=Fd
+        )
+        return res
+
+    # In all functions, im1 is R (reference, or template) and im2 is N (new, or science)
+    def computeDiffimFourierSpace(self, **kwargs):
+        """Compute ZOGY diffim `D` as proscribed in ZOGY (2016) manuscript
+
+        Compute the ZOGY eqn. (13):
+        $$
+        \widehat{D} = \frac{Fr\widehat{Pr}\widehat{N} -
+        F_n\widehat{Pn}\widehat{R}}{\sqrt{\sigma_n^2 Fr^2
+        |\widehat{Pr}|^2 + \sigma_r^2 F_n^2 |\widehat{Pn}|^2}}
+        $$
+        where $D$ is the optimal difference image, $R$ and $N$ are the
+        reference and "new" image, respectively, $Pr$ and $P_n$ are their
+        PSFs, $Fr$ and $Fn$ are their flux-based zero-points (which we
+        will set to one here), $\sigma_r^2$ and $\sigma_n^2$ are their
+        variance, and $\widehat{D}$ denotes the FT of $D$.
+
+        Returns
+        -------
+        A lsst.pipe.base.Struct containing:
+        - D : 2D numpy.array, the proper image difference
+        - D_var : 2D numpy.array, the variance image for `D`
+        """
+        # Do all in fourier space (needs image-sized PSFs)
+        psf1 = self._padPsfToImageSize(self.im1_psf, self.im1)
+        psf2 = self._padPsfToImageSize(self.im2_psf, self.im2)
+
+        preqs = self.computePrereqs(psf1, psf2, padSize=0)  # already padded the PSFs
+
+        # First do the image
```

Comment:
Ah but they're not the same (I almost miss that) -- the diffim is a subtraction, the variance is an addition.